### PR TITLE
Add user roles

### PR DESCRIPTION
**Pull request by bjkeller**
1. Introduces ability to add user roles for the projects in a study. User management gear now reads input file with study-specific mapping of authorizations to roles that is used to determine how roles should be added.
2. Updated classes to reduce duplicate code in manage centers and studies

### CODE REVIEW

File: gear/user_management/src/python/user_app/run.py [hunk 1]
```diff
@@ -27,14 +28,15 @@ def read_yaml_file(file_bytes: bytes) -> Optional[List[Any]]:
     Returns:
```
Comment (hewanadun) [inline gear/user_management/src/python/user_app/run.py:28]: Update docstring Args

File: gear/user_management/src/python/user_app/main.py [hunk 2]
```diff
@@ -80,45 +97,37 @@ def update_email(*, proxy: FlywheelProxy, user: User, email: str) -> None:
     proxy.set_user_email(user=user, email=email)
 
 
-# pylint: disable=(too-many-locals)
-def run(*, proxy: FlywheelProxy, user_list, skip_list: Set[str]):
+def run(*, proxy: FlywheelProxy, user_list, admin_group: NACCGroup,
+        skip_list: Set[str], authorization_map: AuthMap):
     """Manages users based on user list."""
```
Comment (hewanadun) [inline gear/user_management/src/python/user_app/main.py:103]: Add docstring

File: gear/user_management/src/python/user_app/main.py [hunk 3]
```diff
@@ -25,36 +44,34 @@ def create_user(proxy: FlywheelProxy, user_entry: UserDirectoryEntry) -> User:
     Returns:
       the ID for flywheel User created from the directory entry
```
Comment (hewanadun) [inline gear/user_management/src/python/user_app/main.py:45]: This method returns the User object (not ID)

File: gear/user_management/src/python/user_app/main.py [hunk 4]
```diff
@@ -25,36 +44,34 @@ def create_user(proxy: FlywheelProxy, user_entry: UserDirectoryEntry) -> User:
...
     """
-    user_id = user_entry.credentials['id']
-    new_id = proxy.add_user(
-        User(id=user_id,
-             firstname=user_entry.name['first_name'],
-             lastname=user_entry.name['last_name'],
-             email=user_id))
-    user = proxy.find_user(new_id)
+    new_id = proxy.add_user(create_user(user_entry=user_entry))
+    user = proxy.find_user(user_entry.user_id)
```
Comment (hewanadun) [inline gear/user_management/src/python/user_app/main.py:48]: Do we need to check new_id and user_entry.user_id are the same?

**Review by hewanadun — DISMISSED** [review]
Looks good. 
Few minor changes are noted in the comments.

File: gear/user_management/src/python/user_app/run.py (hunk 1 above)
Comment (bjkeller) [reply gear/user_management/src/python/user_app/run.py:28]: done

File: common/src/python/centers/center_group.py [hunk 5]
```diff
@@ -383,6 +400,67 @@ def __add_project(self, label: str) -> ProjectAdaptor:
...
+
+        Args:
+          user: the user to add
+          authorizations: the authorizations for the user
+        """
+        assert user.id, "requires user has ID"
+
+        portal_info = self.get_portal_info()
+        study_info = portal_info.studies[authorizations.study_id]
+
```
[the remark this answers is not in the document]
Comment (bjkeller) [reply common/src/python/centers/center_group.py:415]: yes, changed line to used get and then logged warning

File: gear/user_management/src/python/user_app/main.py (hunk 3 above)
Comment (bjkeller) [reply gear/user_management/src/python/user_app/main.py:45]: fixed

File: gear/user_management/src/python/user_app/main.py (hunk 4 above)
Comment (bjkeller) [reply gear/user_management/src/python/user_app/main.py:48]: I don't think so. the User created by create_user has the id set to user_entry.user_id and there is nothing in proxy.add_user that should result in a different value. 

are you aware of anything that would indicate this is an issue?

File: gear/user_management/src/python/user_app/main.py (hunk 2 above)
Comment (bjkeller) [reply gear/user_management/src/python/user_app/main.py:103]: done

**Review by hewanadun — APPROVED** [review]

**Review by hewanadun — DISMISSED** [review]

**Pull request merged**